plugin/src/helpers/config.ts

```diff
@@ -108,3 +128,22 @@ export const configureHandlerFunctions = ({ netlifyConfig, publish, ignore = []
     })
   })
 }
+
+/**
+ * Persist NEXT.js custom headers to the Netlify configuration so the headers work with static files
+ *
+ * @param customHeaders - Custom headers defined in the Next.js configuration
+ * @param netlifyHeaders - Existing headers that are already configured in the Netlify configuration
+ */
+export const generateCustomHeaders = (customHeaders: Header[] = [], netlifyHeaders: NetlifyHeaders = []) => {
+  for (const { source, headers } of customHeaders) {
+    netlifyHeaders.push({
+      for: source,
```

Comment:
[pebble] I'm not 100% sure the Next.js path pattern is exactly compatible with ours. Take a look at the equivalent for redirects

Reply:
@ascorbic, based on e.g,

netlify.toml

```
[build]
command = "yarn build"
publish = ".next"

[[plugins]]
package = "@netlify/plugin-nextjs"
```

`headers()` function in next.config.js

```javascript
/** @type {import('next').NextConfig} */
const nextConfig = {
  async headers() {
    return [
      {
        source: '/',
        headers: [
          {
            key: 'x-hello',
            value: 'yo',
          },
          {
            key: 'x-hello2',
            value: 'yo',
          },
        ],
      },
      {
        source: '/:path*',
        headers: [
          {
            key: 'x-hello',
            value: 'yo path star',
          },
          {
            key: 'x-hello2',
            value: 'yo path star again',
          },
        ],
      },
      {
        source: '/yolo',
        headers: [
          {
            key: 'x-turtle',
            value: 'snapping',
          },
        ],
      },
    ]
  },
};

module.exports = nextConfig
```
and the generated routes-manifest.json

```json
...
  "headers": [
    {
      "source": "/",
      "headers": [
        {
          "key": "x-hello",
          "value": "yo"
        },
        {
          "key": "x-hello2",
          "value": "yo"
        }
      ],
      "regex": "^/(?:/)?$"
    },
    {
      "source": "/:path*",
      "headers": [
        {
          "key": "x-hello",
          "value": "yo path star"
        },
        {
          "key": "x-hello2",
          "value": "yo path star again"
        }
      ],
      "regex": "^(?:/((?:[^/]+?)(?:/(?:[^/]+?))*))?(?:/)?$"
    },
    {
      "source": "/yolo",
      "headers": [
        {
          "key": "x-turtle",
          "value": "snapping"
        }
      ],
      "regex": "^/yolo(?:/)?$"
    }
  ],
  ...
  ```

on post-build, the Netlify configuration is updated to

```
────────────────────────────────────────────────────────────────
  5. @netlify/plugin-nextjs (onPostBuild event)                 
────────────────────────────────────────────────────────────────

Next.js cache saved.
Netlify configuration property "headers" value changed to [
  { for: '/', values: { 'x-hello': 'yo', 'x-hello2': 'yo' } },
  {
    for: '/:path*',
    values: { 'x-hello': 'yo path star', 'x-hello2': 'yo path star again' }
  },
  { for: '/yolo', values: { 'x-turtle': 'snapping' } }
].
```

If they happen to have redirects in the netlify.toml already, on post-build appends to them

```
────────────────────────────────────────────────────────────────
  5. @netlify/plugin-nextjs (onPostBuild event)                 
────────────────────────────────────────────────────────────────

Next.js cache saved.
Netlify configuration property "headers" value changed to [
  { for: '/*', values: { 'Some-Key': 'Some Value' } },
  { for: '/', values: { 'Some-Key2': 'Some Value Just Slash' } },
  { for: '/', values: { 'x-hello': 'yo', 'x-hello2': 'yo' } },
  {
    for: '/:path*',
    values: { 'x-hello': 'yo path star', 'x-hello2': 'yo path star again' }
  },
  { for: '/yolo', values: { 'x-turtle': 'snapping' } }
].
```